finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala

```diff
@@ -34,11 +34,14 @@ trait Stream {
   def onEnd: Future[Unit]
 
   /**
-   * Wraps this [[Stream]] with a [[StreamProxy]] that calls the provided function on each frame after [[read()]]
+   * Wraps this [[Stream]] with a [[StreamOnFrame]] that calls the provided function on each frame after [[read()]]
+   *
    * @param onFrame the function to call on each frame.
-   * @return a [[StreamProxy]] wrapping this [[Stream]]
+   * @return a [[StreamOnFrame]] wrapping this [[Stream]]
    */
-  def onFrame(onFrame: Try[Frame] => Unit): Stream = new StreamProxy(this, onFrame)
+  def onFrame(onFrame: Try[Frame] => Unit): Stream = new StreamOnFrame(this, onFrame)
+
+  def flatMap(f: Try[Frame] => Seq[Frame]): Stream = new StreamFlatMap(this, f)
```

Comment:
Let's add some scaladoc here including the restrictions on `f` (namely that it has to give back the input frame in the output Seq or take ownership of releasing it)